FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/ApiClientRepository.kt

```diff
@@ -0,0 +1,387 @@
+import com.flowcrypt.email.api.retrofit.response.attester.SubmitPubKeyResponse
+import com.flowcrypt.email.api.retrofit.response.attester.WelcomeMessageResponse
+import com.flowcrypt.email.api.retrofit.response.base.Result
+import com.flowcrypt.email.api.retrofit.response.model.ClientConfiguration
+import com.flowcrypt.email.api.retrofit.response.oauth2.MicrosoftOAuth2TokenResponse
+import com.flowcrypt.email.api.wkd.WkdClient
+import com.flowcrypt.email.extensions.kotlin.isValidEmail
+import com.flowcrypt.email.extensions.org.bouncycastle.openpgp.armor
+import com.google.gson.GsonBuilder
+import com.google.gson.JsonObject
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.MultipartBody
+import okhttp3.OkHttpClient
+import okhttp3.RequestBody.Companion.toRequestBody
+import okhttp3.ResponseBody.Companion.toResponseBody
+import org.pgpainless.algorithm.EncryptionPurpose
+import org.pgpainless.key.info.KeyRingInfo
+import retrofit2.Response
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import java.net.HttpURLConnection
+import java.util.concurrent.TimeUnit
+
+/**
+ * This class describes all network requests over the app.
+ *
+ * @author Denis Bondarenko
+ *         Date: 10/24/19
+ *         Time: 6:10 PM
+ *         E-mail: DenBond7@gmail.com
+ */
+object ApiClientRepository : BaseApiRepository {
+  object FES : BaseApiRepository {
+    /**
+     * @param context Interface to global information about an application environment.
+     * @param baseFesUrlPath A base FES URL path.
+     * @param domain A company domain.
+     * @param idToken OIDC token.
+     */
+    suspend fun getClientConfigurationFromFes(
+      context: Context,
+      idToken: String,
+      baseFesUrlPath: String,
+      domain: String,
+    ): Result<ClientConfigurationResponse> =
+      withContext(Dispatchers.IO) {
+        val retrofitApiService = ApiHelper.createRetrofitApiService(context)
+        getResult(context = context) {
+          retrofitApiService.getClientConfigurationFromFes(
+            authorization = "Bearer $idToken",
+            baseFesUrlPath = baseFesUrlPath,
+            domain = domain
+          )
+        }
+      }
+
+    /**
+     * Grab a reply token before uploading a password protected message
+     *
+     * @param context Interface to global information about an application environment.
+     * @param idToken OIDC token.
+     * @param baseFesUrlPath A base FES URL path.
+     */
+    suspend fun getReplyTokenForPasswordProtectedMsg(
+      context: Context,
+      idToken: String,
+      baseFesUrlPath: String,
+    ): Result<MessageReplyTokenResponse> =
+      withContext(Dispatchers.IO) {
+        val retrofitApiService = ApiHelper.createRetrofitApiService(context)
+        getResult(
+          context = context,
+          expectedResultClass = MessageReplyTokenResponse::class.java
+        ) {
```

Comment:
Why here we have to add `expectedResultClass` but for `getClientConfigurationFromFes` we don't have to?

Reply:
It depends on the API error format. unfortunately, not all APIs have the same error format.

Reply:
I've created a separate issue